# Technical documentation

## Outline

#### [Overview](https://github.com/BCODMO/bcodmont/blob/main/docs/technical-documentation.md#overview-1)

#### [New Term Requests](https://github.com/BCODMO/bcodmont/blob/main/docs/technical-documentation.md#new-term-requests-1)

#### [Compiling BCO-SM](https://github.com/BCODMO/bcodmont/blob/main/docs/technical-documentation.md#compiling-bco-sm-1)

#### [Maintaining BCO-SM](https://github.com/BCODMO/bcodmont/blob/main/docs/technical-documentation.md#maintaining-bco-sm-1)

#### [Future Sections](https://github.com/BCODMO/bcodmont/blob/main/docs/technical-documentation.md#future-sections-1)

## Overview

BCODMONT is an OBO style application ontology created using the [Ontology Development Kit \(ODK\)](https://github.com/INCATools/ontology-development-kit). As such it makes use of a variety of features generated by ODK such as ontology imports, a standardized makefile, and githhub version release-generation.

The **BCO**DMO **S**emantic **M**odel \(**BCO-SM**\) is created within BCODMONT and acts as a controlled vocabulary for the BCODMO data managers to annotate dataset metadata within a semantic layer to be incorporated into the BCO-DMO Knowledge Graph.

The BCO-SM is loosely based on the [NERC Vocabulary Server \(NERC-VS\)](https://www.bodc.ac.uk/resources/products/web_services/vocab/) model, but draws from terminology from various ontologies within the [Open Biological and Biomedical Ontology \(OBO\) Foundry](http://www.obofoundry.org/). It is based on the clarity of the NERC semantic model structure for describing marine data, while leveraging the diversity, richness, and semantic maturity of the OBO collection.

### Modules:

BCO-SM has the following domain specific modules which can be used together in combination to annotate datasets.

**Figure 1**: Overview of the BCO-SM

![](.gitbook/assets/image%20%282%29.png)

#### Biology

The Biology module consists of several sub-modules [Anatomy](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/biology/anatomy.owl),[ Physiology](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/biology/physiology.owl), Ecology, Biomolecules, and Organism. It is intended to represent the biologically relevant information pertinent to datasets.

#### Chemistry

The Chemistry module consists of two sub-modules [Compound](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/chemistry/compound.owl) and [Element](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/chemistry/element.owl). The Compound sub-module is comprised of a set of chemicals focusing on multi-element compounds, e.g, `nitrate` or `carbohydrate`. Whereas the Element sub-module consists of molecules which are comprised of a single element, including isotopes and valence states, e.g. `elemental Nitrogen` or `Nitrogen-15`.

#### Matrix

The Matrix module represents the environmental context from which datasets were collected. It is based on both the NERC Matrix vocabulary as well as the [MIxS](https://gensc.org/mixs/) \(Minimum Information about any "x" Sequence\) standards from the Genomic Standards Consortium \(GSC\). BCO-SM's Matrix is divided into four sub-modules which like the MIxS model represent the environmental context data at increasing levels of granularity. The first module [Region](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/matrix/region.owl), describes at the highest level if data were collected from a `polar`, `temperate` or `tropical` region.

The next three sub-modules draws from terminology from the Environment Ontology \(ENVO\); following the MIxS ENVO annotation framework which includes the use of \(one or more terms\) from each of the, [Biome](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/matrix/biome.owl), [Context](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/matrix/context.owl), and [Material](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/matrix/material.owl) sub-modules. These are analogous to the `broad-scale environmental context`, `local environmental context`, and `environmental medium`, fields used in the MIxS model. In the **Biome** module are terms by which to report which major environmental system from which a sample originated, namely an `aquatic`, `marine`, or `terrestrial` biome. `Aquatic`referring to both freshwater as well as non-marine saline environments, e.g., saline lakes.

The **Context** sub-module contains terms by which to describe the local vicinity of a sample which may have a causal influences on data, e.g. if the data were collected from an `esturay` or a water layer such as the `marine benthic zone`.

Finally, the **Material** sub-module contains terms for environmental or other material or materials that immediately surrounded or comprised samples e.g., `sea water`, `plankton`, or `sediment`.

#### Operational

The [Operational](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/operational/operational.owl) module describes information about sample processing e.g., `filter fractionation size`, sample post-processing e.g., `external repository` as well as spatiotemporal sample collection information e.g., `latitude` or `collection date`.

#### Physics

The Physics module is split up into two sub-modules. The first of which, [Characteristic](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/physics/characteristic.owl) consists of terms describing qualities, characteristics or attributes of materials \(both biotic and abiotic\), e.g., the `depth` or `salinity` measured in sea water, or the `color` used to describe an organism or water body. The [Phenomenon](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/physics/phenomenon.owl) module consists of physical phenomena either measured or believed to have a causal influence on a sample, e.g., `Taylor column`, `deep ocean current` , or `thunderstorm`. Note at the moment there are some overlapping terms between the Physics Phenomenon and Matrix Context modules. This is intentional but we may want to constrain terms like `epilimnion` to only Matrix.

#### Quantifiers

The [Quantifiers](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/quantifiers/quantifiers.owl) module consists of terms to describe quantitative measurements conducted on samples within datasets, including classes like `concentration`, `length`, or `standard deviation`. Which are to be used to described Entities in conjunction with terms from other modules.

### OBO Ontologies Used

BCODMONT incorporates terms from pre-existing terms within OBO ontologies in an effort to facilitate linkage with outside data repositories as well as to minimize the burden for creation of new richly defined terms. Cross reference the [OBO Foundry hompage](http://www.obofoundry.org/).

This section is split into two subsections the first consists of OBO foundry ontologies which 1\) are guarantied to be imported using the standard Robot import workflow \(from the ODK\), 2\) which contain standardized annotation properties and axioms \(so we know we can reuse them without writing extra conditions in our search or GUI visualization workflows\) and 3\) Have been vetted as following the OBO principal of orthogonality, that is describing a specific domain or set of domains but not overlapping with the information represented by another ontology. The second section describes other ontologies which are referenced but not directly used as imports. There are several reasons for not directly using some ontologies as imports and optionally mapping to others. Non OBO ontologies are not hosted in the same way as OBO ones and aren't guarantied to work with the import workflow, they are also not guarantied to use the same annotation properties \(e.g., definitions\). Some OBO projects maybe deprecated or are not themselves examples of best practices or good sources to draw from. Some non OBO projects, \(e.g. SIO\) are well made and although not guarantied to have the required properties, they are used by other systems. hence mapping to them helps to bolster overall inter-project interoperability, and future efforts to harmonize them with OBO projects.

**The Basic Formal Ontology \(BFO\)**: is the upper level ontology upon which OBO Foundry ontologies are built. BCO-SM makes use of BFO terms like `site`. Browse BFO from [here](https://www.ebi.ac.uk/ols/ontologies/bfo).

**The Chemical Entities of Biological Interest \(CHEBI\)**: is a structured classification of molecular entities of biological interest focusing on 'small' chemical compounds. CHEBI terms comprise most of the BCO-SM chemistry terms e.g. `nitrate` and are also used in other BCO-SM modules such as the Biomolecules submodule from Biology for terms like `DNA` or `RNA`. Browse CHEBI from [here](https://www.ebi.ac.uk/ols/ontologies/chebi).

**The Cell Ontology \(CL\)**: is a structured controlled vocabulary for cell types in animals. BCO-SM makes use of CL terms, e.g. `red blood cell` within the Biology Anatomy module. Browse CL from [here](https://www.ebi.ac.uk/ols/ontologies/CL).

**The Core Ontology for Biology and Biomedicine \(COB\)**: brings together key terms from a wide range of OBO projects to improve interoperability, attempting to standardize terms important across disciplines and ontologies. BCO-SM when possible tries to make use of COB terms within Biology modules, e.g. `organism`, or `cell`, which might also be found in other OBO ontologies. Browse COB from [here](https://www.ebi.ac.uk/ols/ontologies/cob).

**The Environment Ontology \(ENVO\)**: describes environmental features, biomes, ecosystems, planetary or other astronomical bodies, their parts, or environmental processes. The BCO-SM primarily uses ENVO terms, \(e.g., `abyssal plain` or `first year ice` within the Matrix modules to comprehensively describe the environmental context of collected samples, at various levels of granularity, following the MIxS model used by genomic repositories such as NCBI and EBI. BCO-SM also uses ENVO terms in the Physics Phenomenon and Biology Ecology modules. Browse ENVO from [here](https://www.ebi.ac.uk/ols/ontologies/envo). Additional documentation about how to browse the Environment Ontology \(ENVO\), as well as its contents, can be found at the [navigating ENVO](https://github.com/EnvironmentOntology/envo/wiki/Navigating-ENVO) github wiki page. The navigation page describes how to go about viewing ENVO \(or other OBO ontologies\), as well as described the main hierarchies within ENVO, giving examples of each.

**The Gene Ontology \(GO\)**: describes the function of genes and gene products. BCO-SM uses GO terms, e.g., `photosynthesis` and `repsiration` in the Biology Physiology submodule. Browse GO from [here](https://www.ebi.ac.uk/ols/ontologies/go).

**The Information Artifact Ontology \(IAO\)**: sometimes refereed to as the OBO Metadata Ontology \(OMO\), is the OBO ontology for information entities. IAO includies both classes, e.g., `data set` or `setting datum` and annotation properties, e.g. `defintion`. When Possible IAO classes are used within the BCO-SM Operational module. Browse IAO from [here](https://www.ebi.ac.uk/ols/ontologies/iao).

**The Mass spectrometry ontology \(MS\)**: is a structured controlled vocabulary for the annotation of experiments concerned with proteomics mass spectrometry. Browse MS from [here](https://www.ebi.ac.uk/ols/ontologies/ms).

**The Ontology for Biomedical Investigations \(OBI\)**: is an integrated ontology for the description of life-science and clinical investigations. BCO-SM makes use of various OBI classes, e.g., `sample` and `longitude` within the BCO-SM Operational module. Browse OBI from [here](https://www.ebi.ac.uk/ols/ontologies/obi).

**The Phenotype And Trait Ontology \(PATO\)**: is an ontology of phenotypic qualities, aka, properties, attributes or characteristics. These include both biological attributes of organisms, e.g. `size` as well as physical qualities such as `salinity`, or `acidity`, or `color`. BCO-SM uses PATO in both the Quantifiers and Physics Characteristic modules. Browse PATO from [here](https://www.ebi.ac.uk/ols/ontologies/pato).

**The Population and Community Ontology \(PCO\)**: is an ontology about groups of interacting organisms such as populations and communities. PCO terms such as `Simpson diversity index`, and `community` are used in the Biology Ecology module. Browse PCO from [here](https://www.ebi.ac.uk/ols/ontologies/pco).

**The Relation Ontology \(RO\)**: is a controlled set of relationship types shared across OBO ontologies in order to enable interoperability across ontologies. The BCO-SM module currently do not heavily leverage RO relationships, however future iterations of the system could to enable more advanced querying. Browse RO from [here](https://www.ebi.ac.uk/ols/ontologies/ro).

**The Statistical Methods Ontology \(STATO\)**: is a general-purpose statistics Ontology, providing coverage for processes such as statistical tests, their conditions of application, and information needed or resulting from statistical methods. BCO-SM leverages STATO terms such as `median` , or `confidence interval` in the Quantifiers module. Browse STATO from [here](https://www.ebi.ac.uk/ols/ontologies/stato).

**The multi-species anatomy ontology \(UBERON\)**: is an integrated cross-species anatomy ontology covering animals and bridging multiple species-specific ontologies. BCO-SM uses UBERON terms such as `blood`, or `juvenile` in Biology modules such as Anatomy and Physiology respectively. Browse UBERON from [here](https://www.ebi.ac.uk/ols/ontologies/uberon).

**The Sequence types and features ontology \(SO\)**: is a structured controlled vocabulary for sequence annotation, for the exchange of annotation data and for the description of sequence objects in databases. Browse SO from [here](https://www.ebi.ac.uk/ols/ontologies/SO). //TODO will need to add this as new import \(can use as example on how to add new ones\).

**The NCBI organismal classification \(NCBITaxon\)**: is An ontology representation of the [NCBI organismal taxonomy database](https://www.ncbi.nlm.nih.gov/taxonomy), which is in turn a curated classification and nomenclature for all of the organisms in the public sequence databases. Not yet used but will likely be picked up for use in the Biology Organism module.

### Other Ontologies Referenced but not directly used

**The chemical methods ontology \(CHMO\)**: describes methods used to collect data in chemical experiments. CHMO imports from various other ontologies and is still TBD if it provides sufficient terminology to merit inclusion in BCO-SM.

**The Ontology of Bioscientific Data Analysis \(**[**EDAM**](https://edamontology.org/page)**\)**: is a representation of concepts that are prevalent within computational biology, bioinformatics, and bioimage informatics. including types of data and data identifiers, data formats, operations, and topics related to data analysis in life sciences. EDAM is **NOT** part of the OBO foundry so doesn't conform the the shared design principles or use of shared annotation properties. Although some BCO-SM terminology is based on and mapped to EDAM, it's not officially used or imported into BCO-SM.

**The Genomic Epidemiology Ontology \(GENEPIO\)**: covers concepts relevant to foodborne pathogens and associated outbreaks. A few GENEPIO terms maybe relevant to BCO-SM however, it's TBD if it will be included.

**The NCI Thesaurus OBO Edition \(NCIT\)**: is an ontology style \(owl\) representation of the National Cancer Institute \([NCI](https://ncicb.nci.nih.gov/)\) thesaurus. NCIT includes broad coverage of the cancer domain, including cancer related diseases, findings and abnormalities. NCIT is not an ontology of concepts merely a representation of a thesaurus as an ontology. Although, NCIT has a broad coverage of terminology and is reasonable to cross-reference, it should not be considered as an an ontological framework and hence is not imported into BCO-SM. Importing NCIT is probably not worth the effort to maintain it as a commutable structure if it's not organized to be such a product.

**The NMR-instrument specific component of metabolomics investigations Ontology \(NMR\)**: was previously used to describe experimental conditions of the Nuclear Magnetic Resonance \(NMR\) component in a metabolomics investigation. However it is now **deprecated** by the OBO foundry. Hence BCO-SM shouldn't use it.

**The Ontologized MIABIS \(OMIABIS\)**: is an ontological version of MIABIS \(Minimum Information About BIobank data Sharing\). MIABIS aims to provide common biobank terminology to facilitate data sharing in biobanks and sample collections. Some terminology from OMIABIS maybe of use to BCO-SM TBD.

**The PRIDE PRoteomics IDEntifications \(PRIDE\)**: is a controlled vocabulary used in the [PRIDE database](https://www.ebi.ac.uk/pride/markdownpage/citationpage) for proteomics data, including protein and peptide identifications, post-translational modifications and supporting spectral evidence. PRIDE is **not** part of the OBO foundry so can't be guaranteed to be imported using the standard import workflow, or interoperable with the other vocabularies. Some pride terms might be useful TBD if we cross-reference it.

**The Semanticscience Integrated Ontology \(SIO\)**: is an ontology providing descriptions of objects, processes and their attributes. SIO is **not** part of the OBO foundry so can't be guaranteed to be imported using the standard import workflow, or interoperable with the other vocabularies. There exist many SIO terms which might be useful to BCO-SM hence will will cross reference it, for example in the Biology Biomolecules module.

## New Term Requests

In the interim stages of the BCO-SM's development we expect there to be many changes made to the model, mostly additions but perhaps also some term replacements or reshuffling. As such this section serves as a guide for data managers to request new terms to be added to BCO-SM. Once the model is a bit more solidified and the downstream data processing in order, the process of adding new terms could be taken up by interested Data Mangers.

For the time being, data managers can do the following to request new terms.

1\) First navigate to the [bcodmont/issues](https://github.com/BCODMO/bcodmont/issues/new) page.

2\) Create a new issue describing the requested term. Explain what it is in clear concise language what the missing term is, what BCO-SM module you think it should be added to, and where in the module it should be located \(with respect to pre-existing terms\). Make sure for now to tag @kaiiam in the github issue \(until the model is stabilized and others are trained to handle this as well or in place\).

3\) Make sure to thoroughly check through the existing modules in [https://github.com/BCODMO/bcodmont/tree/main/src/ontology/BCODMO\_SM](https://github.com/BCODMO/bcodmont/tree/main/src/ontology/BCODMO_SM). //TODO either add link to web protege versions or other way to visualize these such as the EBI-docker.

4\) Next, search for appropriate pre-existing terms in OBO ontologies. Note that when possible BCO-SM uses existing OBO terms to both have greater interoperability with external projects as well as reuse the work of countless people who have contributed to many well-grounded biological and biomedical ontologies. To explore existing OBO ontology terms navigate to the [OLS ontology lookup page](https://www.ebi.ac.uk/ols/index) and type in your term into the search bar. There are many OBO ontologies; however, not all of which are of the same level of quality, therefore it can be a bit confusing to search. See the [OBO Ontologies Used](technical-documentation.md#obo-ontologies-used) section of [Overview](technical-documentation.md#overview) for a list of the OBO ontologies we are using for BCO-SM. This list is by no means the only OBO ontologies to search for terms in, but it represents a very good place to start.

5\) **Optional** if you want to go the extra mile, check out the following section on preparing new BCO-SM terms for the step-by-step instructions for adding to modules via the Robot templates by which they are compiled. Make sure to check that if any new import import terms are required \(aka if we can leverage existing ontology terms\). If so, make sure to add them to the list of imports, see the [Managing Imports](https://github.com/BCODMO/bcodmont/blob/main/docs/technical-documentation.md#managing-imports) section prior to [compiling](https://github.com/BCODMO/bcodmont/blob/main/docs/technical-documentation.md#compiling-bco-sm-1) the robot templates. Also note that when using imported terms, we do **NOT** need to add annotion properties such as synonyms or definitions when filling out the robot templates in the following section.

### Preparing new BCO-SM terms

The working versions of the BCO-SM robot templates are currently the google drive sheets \(which are also linked within the _BCO-DMO Semantic Data Project_ trello page\).  The links are to the working BCO-SM Robot Templates are as follows: [Biology vocab](https://docs.google.com/spreadsheets/d/11Cdr6OMGnJefhctlowwYL9YMHg8qiZglMJlTPH9mkuI/edit?usp=sharing), [Chemistry vocab](https://docs.google.com/spreadsheets/d/1_yTKHLinTjgAUSCfgXGInGKrWWvehJbK0CrVDvwfd54/edit?usp=sharing), [Qualifiers vocab](https://docs.google.com/spreadsheets/d/1I-M7Z9qrtTt4YdfwnU86zYzsQt3lgYPiegyXAjHZSlc/edit?usp=sharing), [Physics vocab](https://docs.google.com/spreadsheets/d/1H_iNCD3LzZ9LflzZXUDTU5UFtwfi594mbJRZqmUNYs4/edit?usp=sharing), [Matrix vocab](https://docs.google.com/spreadsheets/d/12vG3U698AEc7t31bffGhzOWybqIXJOP5Rc6eGsbr1I0/edit?usp=sharing), [Operational vocab](https://docs.google.com/spreadsheets/d/1OoKDrWMdFZa0oiGSmJe1rYcKVuWOoqa2evEkIP_wxew/edit?usp=sharing).

At the moment these are exported down to the [BCODMO\_SM](https://github.com/BCODMO/bcodmont/tree/main/src/ontology/BCODMO_SM) sub-directories e.g., [biology/robot\_templates/anatomy.tsv](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/biology/robot_templates/anatomy.tsv). However, relying on google sheets as the source of truth should not to be considered a long term solution. At some point, once we've stabilized the system, we will switch to treating the `tsv` files within the `X_module/robot_templates/` to being the source of truth \(place to work from\).

#### Basics about Robot templates

Robot templates are structured as follows, the first row contains `headers`, these are human-readable labels that are not compiled by robot. These can be renamed as needed. The second row contains the `templates` which must contain strings formatted as specified in the [Robot template documentation](https://robot.obolibrary.org/template). Please do not change these. All subsequent rows are for the addition of new terms or the modification of existing ones.

#### BCO-SM Category Headers

At the moment the first several columns contain headers such as `Sub category 1:`. These columns are not complied by the robot templates but are used to visualize the basic parent child hierarchy that is complied when the template is converted to owl. These are not strictly necessary \(and could be removed, however, they can be helpful when adding new terms to the model\).

#### Ontology IDs

In order for a term \(a row in the spreadsheet\) to be recognized by robot and compiled into owl code, it must have an entry in the `ontology ID` column. This must either include a BCODMO Semantic Model \(BCO-SM\) ID, or an ID from another OBO ontology. BCO-SM terms must start with the prefix `BSM:` and be followed by seven digits e.g., `BSM:0200001`. BCO-SM IDs must be within an appropriate numeric ID range as specified within the [BCODMO\_SM/idranges.owl](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/idranges.owl) file.

* Quantifiers: 0000001-0010000
* Biology:
  * Anatomy: 0010001-0020000
  * Physiology: 0020001-0030000
  * Ecology: 0030001-0040000
  * Biomolecules: 0040001-0050000
  * Organism: 0050001-0060000
* Matrix:
  * Material: 0100001-0110000
  * Context: 0110001-0120000
  * Biome: 0120001-0130000
* Chemistry:
  * Element: 0200001-0210000
  * Compound: 0210001-0220000
* Physics:
  * Characteristic: 0310001-0320000
  * Phenomenon: 0320001-0330000
* Operational: 0400001-0410000

Imported terms from other OBO ontologies such as ENVO, UBERON, GO etc must follow this same formatting, e.g, with imported ENVO terms one must use the prefix `ENVO:` followed by the correct identifier number, e.g. `ENVO:03000102`. This is refereed to as the `curie` ID formatting style.

#### Labels

In the `label` column, put the primary labels for new or imported terms, e.g., `marsh` or `ice`. BCO-SM follows lower case naming conventions, with exceptions for proper nouns, e.g., `Taylor column` or `WMO blizzard`. In some instances BCO-SM renames terms imported from existing OBO ontologies in order to simplify them. For example, we rename `ENVO:01000811` `nitrogen-oxygen planetary atmosphere` to `atmosphere`.

#### Parent classes

BCO-SM terms and imported terms should make use of the `parent class` relationship. An important part of the work of an ontologist is to identify parent/child \(super/sub\) class relationships. For example if you were adding a new class `ice fog`, you'd add the pre-existing parameter `fog` as the `parent class` for the new term. For the most part BCO-SM follows the hierarchical relationships that already exist within imported OBO ontologies. However in some instances, e.g., Chemistry we have simplified the hierarchy, and removed some intermediate terms. If adding new terms which themselves don't have an appropriate parent class in BCO-SM, it might be necessary to add intermediate parent classes. See [this issue](https://github.com/BCODMO/bcodmont/issues/4) as an example of doing so. 

In order to add to a parent class to a Robot Template, you can simply type or copy paste the label of an existing term into the `parent class` column.

To discover new OBO terms which might be relevant to use as parent classes, you can browse them from the ENVO from the [EMBL-EBI Ontology Lookup Service](https://www.ebi.ac.uk/ols/ontologies/) \(OLS\). The OLS search can be filtered by ontology, e.g., [https://www.ebi.ac.uk/ols/ontologies/chebi](https://www.ebi.ac.uk/ols/ontologies/chebi), or [https://www.ebi.ac.uk/ols/ontologies/go](https://www.ebi.ac.uk/ols/ontologies/go) etc.

#### Matches

BCO-SM makes use of the SKOS `exact match` and `close match` annotation properties. `exact match` should be used in cases where two semantics are intended to be the same. For example if an OBO term is found or created that can replace a BCO-SM term.`close match` can be used in cases where two semantics may be similar but aren't exactly the same representation. For example terms form the [NCI Thesaurus OBO Edition](https://www.ebi.ac.uk/ols/ontologies/ncit) e.g., `NCIT:C12508`.

#### Definitions

Although definitions and other annotation properties are included for imported OBO terms, new BCO-SM terms, are envisioned to eventually be added to appropriate OBO Foundry Ontologies. As such, we encourage new BCO-SM terms to make use of OBO conventions for writing definitions. This includes advising people to follow a genus-differentia structure, i.e., `An A is a B that C's`. The genus refers the parent term, and the differentia is that which makes a term different \(or more specific\) than its parent. Doing this also helps to identify parent/child relationships between terms.

For example in ENVO, they define [land ice mass](http://purl.obolibrary.org/obo/ENVO_01001547) as:

> An ice mass which has formed over land.

Another example, in UBERON, they define [tissue](http://purl.obolibrary.org/obo/UBERON_0000479) as:

> \[A\] multicellular anatomical structure that consists of many cells of one or a few types, arranged in an extracellular matrix such that their long-range organisation is at least partly a repetition of their short-range organisation.

Using a genus-differentia structure helps both to formulate a concise and simple definition as well as determine the appropriate parent class. In cases where there are multiple differentiating factors, they can be written out as a list, for example:

> A blizzard which has 1\) winds of 35 miles per hour \(approximately 56 kilometers per hour\), 2\) lasts for a period of at least three hours, and 3\) occurs in extremely cold temperatures.

For more information about writing ontology definitions please see [this blog post](https://douroucouli.wordpress.com/2019/07/08/ontotip-write-simple-concise-clear-operational-textual-definitions/) written by Dr. Chris Mungall.

#### Comments

In the `comment` column add any supporting information which is relevant to the term but not essential for its definition. As described in the above [blog post](https://douroucouli.wordpress.com/2019/07/08/ontotip-write-simple-concise-clear-operational-textual-definitions/), any information which gets cut out while trying to write a concise genus-differentia style definition can be put into the comment.

#### Editors notes

The `editors note` column serves as a place for ontology developers to make notes about how to better engineer the term or discuss future developments. For new BCO-SM terms this is where one can suggest an an appropriate OBO ontology and possible parent class for such a term.

#### Synonyms

Following OBO common practices, the BCO-SM robot templates contain four types of synonyms: `exact synonym`, `broad synonym`, `narrow synonym`, `related synonym`. These follow SKOS style \(broader or narrow\) conventions.

Use `exact synonym` for interchangeable term labels, for example `seawater` and `sea water`.

Use `broad synonym` when a synonym might refer to more than just the term of interest, for example the term `acid rainfall` has `acid rain` as a broad synonym because `acid rain` might refer to more than just the process of acid rainfall.

When a synonym refers to something more specific than the class label use `narrow synonym`, for example `road` has narrow synonym `highway`.

If you unsure about a synonym being broader or narrower then use `related synonym`.

#### Cross references

The `cross reference` column provides a generic way to link to other resources such as mappings to other vocabularies. This could, for example, be used to map BCO-SM terms to \(individual\) terms from the [NERC vocabularies](https://www.bodc.ac.uk/resources/products/web_services/vocab/).

#### BCODMO master parameter

The `BCODMO master parameter` column is used to link BCO-SM terms to the \(previous system the\) BCODMO master parameters, e.g., `http://lod.bco-dmo.org/id/parameter/1982`.

#### Subsets

BCO-SM has both `category subset` and `Entity subset` columns. At the moment the `Entity subset` column is not being used, however the `category subset` is essential for the BCO-SM compilation process. As such the `category subset` must include a module-specific iri such as `http://bcodmo/sm#material` or `http://bcodmo/sm#anatomy`.

#### Columns with multiple entries

Any columns with `SPLIT=|` at the end of the string in the template \(second\) row allow for multiple `|` delimited inputs where each of group is compiled separately. For example adding `boulbie|buran|purga` within the `narrow synonym` column, which has the template `AL oboInOwl:hasNarrowSynonym@en SPLIT=|` will produce three separate narrow synonym annotation properties when the template is compiled:

![image](https://user-images.githubusercontent.com/12255688/100488568-008bce00-30dd-11eb-8c83-45324df24f0f.png)

## **Compiling BCO-SM**

**//TODO perhaps add** something about when to commit i.e, with a set of new terms to a module. Original commands can be found in the `bcodmont/src/ontology/BCODMO_SM/robot_commands.md` file.

### 1\) Download working copy of Robot Templates

This step is temporary as it draws from the google drive sheets. Later once the model is more settled, this can be changed to simply modify the `.tsv` files in the github repository. Perhaps we could also use additional/fancier tools such as [COGS](https://github.com/ontodev/cogs) to work with the source files in google sheets but have their working versions stored in github \(TBD\).

Once the new additions have been made to the Robot templates \(following the steps described in the [Preparing new BCO-SM terms](https://github.com/BCODMO/bcodmont/blob/main/docs/technical-documentation.md#preparing-new-bco-sm-terms) section, the working google sheet versions of the templates can be pulled down to the github working `.tsv` files.

To download the working google sheet versions of the vocabularies on the google sheet click `File` Then `Download`, then select `Tab-seperated valuse` makeing sure you're on the correct sheet for the module you intend to be on. See the following image below:

Once downloaded save the file in excel or move it on the command line to replace the appropriate Robot Template module `.tsv`file. For example for the `Chemistry compound` tsv sheet, save or move it to be the file at the location `bcodmont/src/ontology/BCODMO_SM/chemistry/robot_templates/compound.tsv`. It should possible to run this step via the command line `wget` or `rsync` TBD.

**Biology**:

Anatomy: `bcodmont/src/ontology/BCODMO_SM/biology/robot_templates/anatomy.tsv`

**Physiology**: `bcodmont/src/ontology/BCODMO_SM/biology/robot_templates/physiology.tsv`

**Matrix:**

Region: `bcodmont/src/ontology/BCODMO_SM/matrix/robot_templates/region.tsv`

Biome: `bcodmont/src/ontology/BCODMO_SM/matrix/robot_templates/biome.tsv`

Context: `bcodmont/src/ontology/BCODMO_SM/matrix/robot_templates/context.tsv`

Material: `bcodmont/src/ontology/BCODMO_SM/matrix/robot_templates/material.tsv`

**Chemistry**

Element: `bcodmont/src/ontology/BCODMO_SM/chemistry/robot_templates/element.tsv`

Compound: `bcodmont/src/ontology/BCODMO_SM/chemistry/robot_templates/compound.tsv`

**Physics**:

Characteristic: `bcodmont/src/ontology/BCODMO_SM/physics/robot_templates/characteristic.tsv`

Phenomenon: `bcodmont/src/ontology/BCODMO_SM/physics/robot_templates/phenomenon.tsv`

**Operational:** `bcodmont/src/ontology/BCODMO_SM/operational/robot_templates/operational.tsv`

### 2\) Run Robot Templates

After filling out the robot templates \(see previous section\), they need to be compiled into owl modules. The first step of compiling the BCO-SM modules involves running the working Robot `tsv` files. Run the following commands in order to generate the `/robot_templates/*.owl` files. Note that these owl files are not the final module owl, products, they are only an intermediate step on the way toward compiling the final owl files.

These commands make use of the [Robot template command](https://robot.obolibrary.org/template) which compiles owl or other ontology formats based on an input tabular file such as a `.tsv` or `.csv` file. The `--template` flag specifies the input template, here the BCO-SM tsv module files. The `-i` or input flag specifies the input ontology \(and is required\). The `-o` flag designates the output file.

Run the following commands from the `bcodmont/src/ontology/BCODMO_SM/` directory.

#### Biology

Run **anatomy**.tsv:

```text
robot template --template biology/robot_templates/anatomy.tsv -i ../bcodmont-edit.owl --prefix "RO:http://purl.obolibrary.org/obo/RO_" --prefix "COB:http://purl.obolibrary.org/obo/COB_" --prefix "BSM:http://purl.obolibrary.org/obo/BSM_"  --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/biology/robot_templates/anatomy.owl" -o biology/robot_templates/anatomy.owl
```

Run **ecology**.tsv:

```text
robot template --template biology/robot_templates/ecology.tsv -i ../bcodmont-edit.owl --prefix "RO:http://purl.obolibrary.org/obo/RO_" --prefix "ENVO:http://purl.obolibrary.org/obo/ENVO_" --prefix "PCO:http://purl.obolibrary.org/obo/PCO_" --prefix "BSM:http://purl.obolibrary.org/obo/BSM_"  --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/biology/robot_templates/ecology.owl" -o biology/robot_templates/ecology.owl
```

Run **physiology**.tsv:

```text
robot template --template biology/robot_templates/physiology.tsv -i ../bcodmont-edit.owl --prefix "RO:http://purl.obolibrary.org/obo/RO_" --prefix "BSM:http://purl.obolibrary.org/obo/BSM_"  --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/biology/robot_templates/physiology.owl" -o biology/robot_templates/physiology.owl
```

#### Quantifiers

Run **quantifiers**.tsv:

```text
robot template --template quantifiers/robot_templates/quantifiers.tsv -i ../bcodmont-edit.owl --prefix "RO:http://purl.obolibrary.org/obo/RO_" --prefix "BSM:http://purl.obolibrary.org/obo/BSM_" --prefix "SIO:http://semanticscience.org/resource/SIO_" --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/quantifiers/robot_templates/quantifiers.owl" -o quantifiers/robot_templates/quantifiers.owl
```

#### Chemistry

Run **element**.tsv:

```text
robot template --template chemistry/robot_templates/element.tsv -i ../bcodmont-edit.owl --prefix "RO:http://purl.obolibrary.org/obo/RO_" --prefix "BSM:http://purl.obolibrary.org/obo/BSM_" --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/chemistry/robot_templates/element.owl" -o chemistry/robot_templates/element.owl
```

Run **compound**.tsv:

```text
robot template --template chemistry/robot_templates/compound.tsv -i ../bcodmont-edit.owl --prefix "RO:http://purl.obolibrary.org/obo/RO_" --prefix "BSM:http://purl.obolibrary.org/obo/BSM_" --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/chemistry/robot_templates/compound.owl" -o chemistry/robot_templates/compound.owl
```

#### Physics

Run **characteristic**.tsv:

```text
robot template --template physics/robot_templates/characteristic.tsv -i ../bcodmont-edit.owl --prefix "RO:http://purl.obolibrary.org/obo/RO_" --prefix "BSM:http://purl.obolibrary.org/obo/BSM_" --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/physics/robot_templates/characteristic.owl" -o physics/robot_templates/characteristic.owl
```

Run **phenomenon**.tsv:

```text
robot template --template physics/robot_templates/phenomenon.tsv -i ../bcodmont-edit.owl --prefix "RO:http://purl.obolibrary.org/obo/RO_" --prefix "BSM:http://purl.obolibrary.org/obo/BSM_" --prefix "SIO:http://semanticscience.org/resource/SIO_" --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/physics/robot_templates/phenomenon.owl" -o physics/robot_templates/phenomenon.owl
```

#### Matrix

Run **material**.tsv:

```text
robot template --template matrix/robot_templates/material.tsv -i ../bcodmont-edit.owl --prefix "RO:http://purl.obolibrary.org/obo/RO_" --prefix "BSM:http://purl.obolibrary.org/obo/BSM_" --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/matrix/robot_templates/material.owl" -o matrix/robot_templates/material.owl
```

Run **context**.tsv:

```text
robot template --template matrix/robot_templates/context.tsv -i ../bcodmont-edit.owl --prefix "RO:http://purl.obolibrary.org/obo/RO_" --prefix "BSM:http://purl.obolibrary.org/obo/BSM_" --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/matrix/robot_templates/context.owl" -o matrix/robot_templates/context.owl
```

Run **biome**.tsv:

```text
robot template --template matrix/robot_templates/biome.tsv -i ../bcodmont-edit.owl --prefix "RO:http://purl.obolibrary.org/obo/RO_" --prefix "BSM:http://purl.obolibrary.org/obo/BSM_" --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/matrix/robot_templates/biome.owl" -o matrix/robot_templates/biome.owl
```

Run **region**.tsv:

```text
robot template --template matrix/robot_templates/region.tsv -i ../bcodmont-edit.owl --prefix "RO:http://purl.obolibrary.org/obo/RO_" --prefix "BSM:http://purl.obolibrary.org/obo/BSM_" --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/matrix/robot_templates/region.owl" -o matrix/robot_templates/region.owl
```

#### Operational

Run **operational**.tsv:

```text
robot template --template operational/robot_templates/operational.tsv -i ../bcodmont-edit.owl --prefix "RO:http://purl.obolibrary.org/obo/RO_" --prefix "BSM:http://purl.obolibrary.org/obo/BSM_" --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/operational/robot_templates/operational.owl" -o operational/robot_templates/operational.owl
```

### 3\) Merge Robot templates

In section 2 we merge the robot template preliminary outputs with the ontology imports so that they later can be merged together with all desired annotation properties including those from the imports. For example, in our robot templates when we are importing a term from an existing OBO ontology, we don't need to retype it's definition because we will get that from the import version of that ontology \(assume the term is imported properly\). In this step definitions or other annotation properties are joined together with the preliminary owl robot template products. In step 3 we will filter these joint products for only those those terms specified in the original robot template `tsv` files.

As before we can run these commands from `bcodmont/src/ontology/BCODMO_SM/`

#### Merge Imports and Preliminary Robot Templates

The following step merges imports with the robot templates without removing any axioms to produce the `BCODMO_SM_merged.owl` file. This version is used for **all modules other than the two chemistry modules.** Here we are intentionally not removing CHEBI or ENVO axioms. This step makes use of the [Robot merge command](https://robot.obolibrary.org/merge) which merges multiple ontology files into a single file. //TODO "Run this step after modifying any of the following ... imports or ... modules.". Here we are effectively makeing a merged file in which we are **not removing** axioms. Note that some ontologies, e.g., CHEBI, ENVO are used both here and in the next step for the axiom removed ontology merge product. The commands in later steps clarify which merge product is used for what modules. 

```text
robot merge --input ../imports/envo_import.owl --input ../imports/pato_import.owl --input ../imports/uberon_import.owl --input ../imports/go_import.owl --input ../imports/iao_import.owl --input ../imports/obi_import.owl --input ../imports/uo_import.owl --input ../imports/chebi_import.owl --input ../imports/stato_import.owl --input ../imports/ms_import.owl --input ../imports/bfo_import.owl --input ../imports/cl_import.owl --input biology/robot_templates/physiology.owl --input physics/robot_templates/characteristic.owl --input physics/robot_templates/phenomenon.owl --input quantifiers/robot_templates/quantifiers.owl --input operational/robot_templates/operational.owl annotate --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/merge_products/BCODMO_SM_merged.owl" --version-iri "http://purl.obolibrary.org/BCODMO_SM/merge_products/BCODMO_SM_merged.owl" --output merge_products/BCODMO_SM_merged.owl
```

#### Make Object Property \(OP\) Free Versions of Select Import Ontologies

These steps will create more intermediate products consisting of the import ontologies without object properties \(e.g. subclass relations\). These axiom-free versions of the import ontologies will be merged together in the next step to create an axiom free merged product which is used for several modules. This is done as certain BCO-SM modules simplify existing hierarchies within certain OBO ontologies \(e.g, CHEBI\). Doing this allows for a simpler hierarchy to be shown in the final output files. If needed, we can add more with more input ontologies.

**Make OP free version of CHEBI import**. Only need to run after adding to CHEBI import.

```text
robot remove --input ../imports/chebi_import.owl --axioms logical annotate --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/intermediate/chebi_import_axioms_removed.owl" --version-iri "http://purl.obolibrary.org/BCODMO_SM/intermediate/chebi_import_axioms_removed.owl" --output intermediate/chebi_import_axioms_removed.owl
```

**Make OP free version of ENVO import**. Only run after adding to ENVO import.

```text
robot remove --input ../imports/envo_import.owl --axioms logical annotate --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/intermediate/envo_import_axioms_removed.owl" --version-iri "http://purl.obolibrary.org/BCODMO_SM/intermediate/envo_import_axioms_removed.owl" --output intermediate/envo_import_axioms_removed.owl
```

**Make OP free version of UBERON import**. Only run after adding to UBERON import.

```text
robot remove --input ../imports/uberon_import.owl --axioms logical annotate --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/intermediate/uberon_import_axioms_removed.owl" --version-iri "http://purl.obolibrary.org/BCODMO_SM/intermediate/uberon_import_axioms_removed.owl" --output intermediate/uberon_import_axioms_removed.owl
```

**Make OP free version of GO import**. Only run after adding to GO import.

```text
robot remove --input ../imports/go_import.owl --axioms logical annotate --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/intermediate/go_import_axioms_removed.owl" --version-iri "http://purl.obolibrary.org/BCODMO_SM/intermediate/go_import_axioms_removed.owl" --output intermediate/go_import_axioms_removed.owl
```

**Make OP free version of PCO import**. Only run after adding to PCO import.

```text
robot remove --input ../imports/pco_import.owl --axioms logical annotate --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/intermediate/pco_import_axioms_removed.owl" --version-iri "http://purl.obolibrary.org/BCODMO_SM/intermediate/pco_import_axioms_removed.owl" --output intermediate/pco_import_axioms_removed.owl
```

#### Merge Modules with Axiom free Import Ontologies

This step will merge the axiom free import ontologies \(CHEBI, ENVO, UBERON, and GO\) generated in the above steps with select modules in which we are electing not to use the original hierarchies from. These modules include: both Chemistry modules \(**compound** and **element**\), all four **matrix** modules \(region, biome, context and material\), and both current Biology modules \(**anatomy** and **physiology**\).

```text
robot merge --input intermediate/chebi_import_axioms_removed.owl --input intermediate/envo_import_axioms_removed.owl --input intermediate/uberon_import_axioms_removed.owl --input intermediate/go_import_axioms_removed.owl --input ../imports/iao_import.owl --input ../imports/cl_import.owl --input intermediate/pco_import_axioms_removed.owl --input biology/robot_templates/anatomy.owl --input biology/robot_templates/physiology.owl --input chemistry/robot_templates/element.owl --input chemistry/robot_templates/compound.owl --input matrix/robot_templates/material.owl --input matrix/robot_templates/context.owl --input matrix/robot_templates/biome.owl --input matrix/robot_templates/region.owl annotate --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/merge_products/BCODMO_SM_axioms_removed_merged.owl" --version-iri "http://purl.obolibrary.org/BCODMO_SM/merge_products/BCODMO_SM_axioms_removed_merged.owl" --output merge_products/BCODMO_SM_axioms_removed_merged.owl
```

### 4\) Filter merge product to create final modules

The following steps make use of the [Robot filter comand](https://robot.obolibrary.org/filter) in order to produce final export versions of the BCO-SM modules including all terms asserted in the Robot template `tsv` files as well as additional annotation properties and object properties brought in from the import ontologies.

#### Biology

Filter **anatomy** from the axiom-free merged ontology

```text
robot filter --input merge_products/BCODMO_SM_axioms_removed_merged.owl --prefix "bsm:http://bcodmo/sm#" --select "oboInOwl:inSubset=bsm:anatomy" --select annotations --signature true annotate --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/biology/anatomy.owl" --version-iri "http://purl.obolibrary.org/BCODMO_SM/biology/anatomy.owl" --output biology/anatomy.owl
```

Filter **physiology** from the axiom-free merged ontology

```text
robot filter --input merge_products/BCODMO_SM_axioms_removed_merged.owl --prefix "bsm:http://bcodmo/sm#" --select "oboInOwl:inSubset=bsm:physiology" --select annotations --signature true annotate --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/biology/physiology.owl" --version-iri "http://purl.obolibrary.org/BCODMO_SM/biology/physiology.owl" --output biology/physiology.owl
```

Filter **ecology** from the axiom-free merged ontology

```text
robot filter --input merge_products/BCODMO_SM_axioms_removed_merged.owl --prefix "bsm:http://bcodmo/sm#" --select "oboInOwl:inSubset=bsm:ecology" --select annotations --signature true annotate --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/biology/ecology.owl" --version-iri "http://purl.obolibrary.org/BCODMO_SM/biology/ecology.owl" --output biology/ecology.owl
```

#### Quantifiers

Filter **quantifiers** from the regular merged ontology

```text
robot filter --input merge_products/BCODMO_SM_merged.owl --prefix "bsm:http://bcodmo/sm#" --select "oboInOwl:inSubset=bsm:quantifiers" --select annotations  --signature true annotate --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/quantifiers/quantifiers.owl" --version-iri "http://purl.obolibrary.org/BCODMO_SM/quantifiers/quantifiers.owl" --output quantifiers/quantifiers.owl
```

#### Chemistry

Filter **element** from the axiom-free merged ontology

```text
robot filter --input merge_products/BCODMO_SM_axioms_removed_merged.owl --prefix "bsm:http://bcodmo/sm#" --select "oboInOwl:inSubset=bsm:element" --select annotations  --signature true annotate --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/chemistry/element.owl" --version-iri "http://purl.obolibrary.org/BCODMO_SM/chemistry/element.owl" --output chemistry/element.owl
```

Filter **compound** from the axiom-free merged ontology

```text
robot filter --input merge_products/BCODMO_SM_axioms_removed_merged.owl --prefix "bsm:http://bcodmo/sm#" --select "oboInOwl:inSubset=bsm:compound" --select annotations  --signature true annotate --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/chemistry/compound.owl" --version-iri "http://purl.obolibrary.org/BCODMO_SM/chemistry/compound.owl" --output chemistry/compound.owl
```

#### Physics

Filter **characteristic** from the regular merged ontology

```text
robot filter --input merge_products/BCODMO_SM_merged.owl --prefix "bsm:http://bcodmo/sm#" --select "oboInOwl:inSubset=bsm:characteristic" --select annotations --signature true annotate --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/physics/characteristic.owl" --version-iri "http://purl.obolibrary.org/BCODMO_SM/physics/characteristic.owl" --output physics/characteristic.owl
```

Filter **phenomenon** from the regular merged ontology

```text
robot filter --input merge_products/BCODMO_SM_merged.owl --prefix "bsm:http://bcodmo/sm#" --select "oboInOwl:inSubset=bsm:phenomenon" --select annotations --signature true annotate --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/physics/phenomenon.owl" --version-iri "http://purl.obolibrary.org/BCODMO_SM/physics/phenomenon.owl" --output physics/phenomenon.owl
```

#### Matrix

Filter **material** from the axiom-free merged ontology

```text
robot filter --input merge_products/BCODMO_SM_axioms_removed_merged.owl --prefix "bsm:http://bcodmo/sm#" --select "oboInOwl:inSubset=bsm:material" --select annotations  --signature true annotate --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/matrix/material.owl" --version-iri "http://purl.obolibrary.org/BCODMO_SM/matrix/material.owl" --output matrix/material.owl
```

Filter **context** from the axiom-free merged ontology

```text
robot filter --input merge_products/BCODMO_SM_axioms_removed_merged.owl --prefix "bsm:http://bcodmo/sm#" --select "oboInOwl:inSubset=bsm:context" --select annotations  --signature true annotate --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/matrix/context.owl" --version-iri "http://purl.obolibrary.org/BCODMO_SM/matrix/context.owl" --output matrix/context.owl
```

Filter **biome** from the axiom-free merged ontology

```text
robot filter --input merge_products/BCODMO_SM_axioms_removed_merged.owl --prefix "bsm:http://bcodmo/sm#" --select "oboInOwl:inSubset=bsm:biome" --select annotations  --signature true annotate --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/matrix/biome.owl" --version-iri "http://purl.obolibrary.org/BCODMO_SM/matrix/biome.owl" --output matrix/biome.owl
```

Filter **region** from the axiom-free merged ontology

```text
robot filter --input merge_products/BCODMO_SM_axioms_removed_merged.owl --prefix "bsm:http://bcodmo/sm#" --select "oboInOwl:inSubset=bsm:region" --select annotations  --signature true annotate --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/matrix/region.owl" --version-iri "http://purl.obolibrary.org/BCODMO_SM/matrix/region.owl" --output matrix/region.owl
```

#### Operational

Filter **operational** from the regular merged ontology

```text
robot filter --input merge_products/BCODMO_SM_merged.owl --prefix "bsm:http://bcodmo/sm#" --select "oboInOwl:inSubset=bsm:operational" --select annotations  --signature true annotate --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/operational/operational.owl" --version-iri "http://purl.obolibrary.org/BCODMO_SM/operational/operational.owl" --output operational/operational.owl
```

## Maintaining BCO-SM

### Managing Imports

In order to re-use ontology terms from existing OBO ontologies we need to import them into our repository. Currently BCO-SM makes use of various OBO ontologies, see the [OBO Ontologies Used](technical-documentation.md#obo-ontologies-used) section.

#### Add new terms to import

In order to import new terms from ontologies already incorporated into BCO-SM, we need to do the following. Note this step is the same as that descirbed in **step 4\)** of [Importing A New Ontology](https://github.com/BCODMO/bcodmont/blob/main/docs/technical-documentation.md#importing-a-new-ontology).

Add new term\(s\) to import to the ontology imports terms file\(s\), e.g., `chebi_terms.txt`. These files are located in the directory path `bcodmont/src/ontology/imports`. For example to add the following import terms to the BCO-SM CHEBI import, you can modify the `chebi_terms.txt` file using any standard text editor such as atom, vim, nano, emacs, gedit, notepad++, textedit, etc. For example add the following lines to the `BCODMO_SM compound terms:`section:

```text
CHEBI:63299	#carbohydrate derivative
CHEBI:16493	#1-deoxy-D-xylulose 5-phosphate
CHEBI:28681	#N,N'-diacetylchitobiose
```

Note that `#`'s are comments any code after which are not compiled, such as the term labels. Some import xyz\_terms.txt files are sorted alphabetically, this insn't stricly necessary and it just to help keep track of existing imports.  Once imported new terms do not need to be duplicated, i.e., we only ever need to specify `CHEBI:63299` once in the so\_terms.txt file \(even if it's used in multiple modules\). Duplications to term imports won't cause problems so it doesn't hurt if it's done on accident.  Imported terms must follow this format referred to as CURIE format with the uppercase ontology name, e.g. `CHEBI`, followed by the numeric ID `63299`. After the curie it's typical but not strictly required to put the term label e.g., `#carbohydrate derivative`. When adding new terms to be imported make sure to add them like this within the appropriate `xyz_terms.txt` file.

Make sure to commit modified `xyz_terms.txt` file\(s\). with an appropriate commit message such as:

```text
git commit chebi_terms.txt -m 'Add CHEBI:63299, CHEBI:16493 and CHEBI:28681 to be imported'
```

#### Running all imports

We can run all existing imports from `bcodmont/src/ontology` with the following command:

```text
./run.sh make all_imports
```

Note when running on linux systems depending on permissions one may need to add `sudo` in front of the above command.

#### Running individual imports

To re-run an individual ontology \(instead of all imported ontologies\) one can run a command such as the following, where for example the Gene Ontology `go` is the ontology of interest. One would run the following from `bcodmont/src/ontology`: \(note `sudo` may be required\). This `run.sh` script is what initiates the call to the docker with the line:`docker run -v $PWD/../../:/work -w /work/src/ontology -e ROBOT_JAVA_ARGS='' -e JAVA_OPTS='' --rm -ti obolibrary/odkfull "$@"`

```text
./run.sh make imports/go_import.owl
```

### Importing A New Ontology

Ontology import is a process by which we fetch a subset of the lastest versions of any given ontology. Hence any changes to imported terms made upsteam of BCO-SM in the source ontology will be pulled down when imports are run. OBO ontologies are generally aware of this and follow comunity best practices making sure not to make extremely dramatic changes which could affect downstream applications such as BCO-SM. It is possible however that a term could be depricated in which case it a replacment term should be provided. Upstream changes to term labels or definitions or axioms are possible but are usually relatively rare. Re-writing our own versions of term labels helps guard against this.

In order to import a new OBO ontology we will need to do the following steps. The example below describes the process to import the **Sequence types and features ontology \(SO\)** into BCODMONT. This process will work with other OBO foundry ontologies which are available from [`http://purl.obolibrary.org/obo/`](http://purl.obolibrary.org/obo/bfo.owl) followed a valid ontology file e.g. `bfo.owl`.

In the following directory `bcodmont/src/ontology`is the ontology `Makefile`. The `Makefile` is the heart of the BCODMONT application ontology created by the Ontology Development Kit \(ODK\) in the initial release. The `Makefile` contains instructions for managing a variety of tasks including managing ontology imports, and releases. Within the `Makefile` \(on approximately line 75 at the time of creating these docs\), there will be the `IMPORTS` code-block which will resemble the following:

```text
IMPORTS = ro pato envo iao obi uo uberon go stato ms bfo cl cob pco chebi

IMPORT_ROOTS = $(patsubst %, imports/%_import, $(IMPORTS))
IMPORT_OWL_FILES = $(foreach n,$(IMPORT_ROOTS), $(n).owl)
IMPORT_FILES = $(IMPORT_OWL_FILES)
```

**Step 1\)** Add the new target ontology to be imported, \(e.g., **`so`**\) to the list of `IMPORTS`, e.g. `IMPORTS = ro pato envo iao obi uo uberon go stato ms bfo cl cob pco chebi so`. Make sure to add an appropriate commit message such as:

```text
git commit Makefile -m 'Add so import to import target list'
```

**Step 2\)** In the `Mirroring upstream ontologies` section:

```text
# ----------------------------------------
# Mirroring upstream ontologies
# ----------------------------------------
#

IMP=true # Global parameter to bypass import generation
MIR=true # Global parameter to bypass mirror generation

## ONTOLOGY: ro
## Copy of ro is re-downloaded whenever source changes
mirror/ro.trigger: $(SRC)

mirror/ro.owl: mirror/ro.trigger
    @if [ $(MIR) = true ] && [ $(IMP) = true ]; then $(ROBOT) convert -I $(URIBASE)/ro.owl -o $@.tmp.owl && mv $@.tmp.owl $@; fi

.PRECIOUS: mirror/%.owl
```

There are a series of make targets for various ontologies e.g., `ro` shown above. Copy and paste the following at the end of the section, replacing `xyz` with the target ontology name in lower case, e.g., `so`.

```text
## ONTOLOGY: xyz
## Copy of xyz is re-downloaded whenever source changes
mirror/xyz.trigger: $(SRC)

mirror/xyz.owl: mirror/xyz.trigger
    @if [ $(MIR) = true ] && [ $(IMP) = true ]; then $(ROBOT) convert -I $(URIBASE)/xyz.owl -o $@.tmp.owl && mv $@.tmp.owl $@; fi

.PRECIOUS: mirror/%.owl
```

Make sure to add an appropriate commit message such as:

```text
git commit Makefile -m 'Add SO import mirror lines'
```

**Step 3\)** Add import terms file.

In the directory `bcodmont/src/ontology/imports` create a new `xyz_terms.txt` file where `xyz` is replaced by the new ontology name, e.g., `so`. This could be done by copying one of the existing files, e.g., `cp bfo_terms.txt xyz_terms.txt`. For SO the command is

```text
cp bfo_terms.txt so_terms.txt
```

**Step 4\)** Add new term\(s\) to import to the `xyz_terms.txt` file.

In our example we'll add to the following to be the content of the `so_terms.txt` file using any standard text editor such as atom, vim, nano, emacs, gedit, notepad++, textedit, etc depending on what system you're using.

```text
### BCODMO_SM Biology/Biomolecules terms:
SO:0001031 #reverse sequence
```

Note the following, `#`'s are comments which are not compiled. Hence line 1 is not strictly necessary but it helps to keep track of where imported terms were intended to be used. Once imported new terms do not need to be duplicated, i.e., we only ever need to specify `SO:0001031`once in the `so_terms.txt` file \(even if it's used in multiple modules\). Imported terms must follow this format referred to as CURIE format with the uppercase ontology name, e.g. `SO`, followed by the numeric ID `0001031`. After the curie it's typical but not strictly required to put the term label e.g., `#reverse sequence`. When adding new terms to be imported make sure to add them like this within the appropriate `xyz_terms.txt` file.

Make sure to **Add** the new .txt file:

```text
git add so_terms.txt
```

As well as commit it with an appropriate commit message such as:

```text
git commit so_terms.txt -m 'Add term imports file for SO'
```

**Step 5\)** Run make imports:

This could either be done with the generic command to run all imports of all included OBO ontologies:

```text
./run.sh make all_imports
```

Or by running an individual import e.g.,

```text
./run.sh make imports/so_import.owl
```

Note that on linux it might be necessary to run with sudo `sudo ./run.sh make imports/so_import.owl`

This will produce the so import owl file containing desired classes which can then be used in the BCO-SM. Make sure to add the new import ontology file

```text
git add imports/so_import.owl
```

As well as commit it with an appropriate message such as:

```text
git commit imports/so_import.owl -m 'Run make all_imports'
```

Note that in these steps we are importing upstream OBO ontologies into BCODMONT, aka this github repository which was setup by running the Ontology Development Kit and serves as the application ontology upon which BCO-SM is built. These imports include all original axioms and annotation properties from the upstream ontologies. Note that although downstream in BCO-SM we remove some axioms from some ontologies for some modules, here we are importing everything from the upstream OBO ontologies.

\*\*\*\*

### **Adding New BSM Modules**

The following section explains how to add a new module to the BCO-SM. See the [modules](https://github.com/BCODMO/bcodmont/blob/main/docs/technical-documentation.md#modules) section for a description and explanation of the current modules in BCO-SM. Note that this likely won't be very commonplace after the initial establishment of the BCO-SM modules, however, just in case the following section describes how to add a new module to BCO-SM. This demonstration will use the `ecology` module as an example. 

The following should be done in the `bcodmont/src/ontology/BCODMO_SM` directory. 

**Step 1\)** identify which of the top level modules `biology`, `chemistry` , `physics`, `operational`, `quantifiers` or `matrix` that the new module should be added to. In our the example, the new `ecology` module should be added to BCO-SMs`Biology`top level module.

**Step 2\)** Create the base robot template `.tsv` file. Within the top level directory folder, e.g., `bcodmont/src/ontology/BCODMO_SM/biology` the directory structure is laid out as follows. There are `.owl` files for each sub module e.g., `anatomy.owl`, and a `robot_templates/`directory which for each modules has both `.owl` and `.tsv`files. Following this stucture we will need to create all of three of these files. The first of which being the base `.tsv` file in the `robot_templates/`directory. For example with the `ecology` example create 1\) the file `ontology/BCODMO_SM/biology/robot_templates/ecology.tsv`. The new `.tsv`file will be a robot template and will thus need to formatted in a particular way, see the ****[Basics about Robot templates](https://github.com/BCODMO/bcodmont/blob/main/docs/technical-documentation.md#basics-about-robot-templates) section for more information. The simplest way to go about this would be to copy the structure, aka the first two header rows of an existing module. This cold be done by for example copying an existing module with a command like `cp biology/robot_templates/anatomy.tsv biology/robot_templates/ecology.tsv`.

**Step 3\)** Make sure to add a new ontology ID range for the new module updating it in both the doccumntation [Ontology IDs](https://github.com/BCODMO/bcodmont/blob/main/docs/technical-documentation.md#ontology-ids) section, as well as the `bcodmont/src/ontology/BCODMO_SM/idranges.owl` file. For example the `ecology` module \(sub module to `biology` will be assigned the range `0030001-0040000`.

**Step 4\)** The next step is to adjust the header rows to be specific for the new module. Although the majority of the header rows can be the same, make sure to update the `Entity subset` and `category subset` columns this is important as the latter is used in later steps to compile the final version of the `owl` modules. The `Entity subset` should be specific for the top level module for example for `biology` it should be `http://bcodmo/sm#biology` note this is not as important as this is not used in the final modle. The `category subset`, however, is required and will need to be updated, to the name of the module, for example in `ecology` will need to be renamed to `http://bcodmo/sm#ecology`. Make sure every single row in the template has these columns filled out. 

**Step 5\)** Fill out the new module with appropriate terms \(making sure not to duplicate any if the file was copied from an existing module in **step 2**\). This will be a bit more work depending on the content being represented by the new module. See the [Preparing new BCO-SM terms](https://github.com/BCODMO/bcodmont/blob/main/docs/technical-documentation.md#preparing-new-bco-sm-terms) section for details about editing the new robot template `.tsv` file. Although it can't be be prescirbed exactly what the content of a new BCO-SM module might be an important consideration is what terminology from existing OBO ontologies can imported and reused in the new module. See the [**OBO Ontologies Used**](https://github.com/BCODMO/bcodmont/blob/main/docs/technical-documentation.md#obo-ontologies-used) section for a list of possibilities, though note that it is not exaustive. Depending on the content other appropirate OBO projects could be brought in. If using any imports, make sure to see and follow the directions in the [**Importing A New Ontology**](https://github.com/BCODMO/bcodmont/blob/main/docs/technical-documentation.md#importing-a-new-ontology) section. In the `ecology` example we'll make use of terms from the Population and Community Ontology \(PCO\) when possible. For non imported terms make sure to only use BCO-SM terms in the ID range specified in **step 3**. Note that this does not need to be completely finished in order to move on the the next steps. As long as their is at least one "data" row in addition to the two header rows you can return to this after completing the remaining steps. It might be helpful to also include some rows containing imported ontology terms to make sure they are being used and imported correctly.  

**Step 6\)** Adding steps to compile BCO-SM. At some point while working on the previous step \(even with just one term in\) you can move on to this and the following steps. See the [**Compiling BCO-SM**](https://github.com/BCODMO/bcodmont/blob/main/docs/technical-documentation.md#compiling-bco-sm-1) section, to which we will add new robot compilation commands. 

**6.1\)** The first part of this \(is the previous step\) aka adding the new content to the module's robot template tsv file. For example `bcodmont/src/ontology/BCODMO_SM/biology/robot_templates/ecology.tsv`. 

**6.2\)** Next we'll need to add a command to run the new robot template see [Run Robot Template](https://github.com/BCODMO/bcodmont/blob/main/docs/technical-documentation.md#2-run-robot-templates). The example command added to comple the ecology template is:

```text
robot template --template biology/robot_templates/ecology.tsv -i ../bcodmont-edit.owl --prefix "RO:http://purl.obolibrary.org/obo/RO_" --prefix "ENVO:http://purl.obolibrary.org/obo/ENVO_" --prefix "PCO:http://purl.obolibrary.org/obo/PCO_" --prefix "BSM:http://purl.obolibrary.org/obo/BSM_"  --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/biology/robot_templates/ecology.owl" -o biology/robot_templates/ecology.owl
```

Note following about this command. the `--template` flag is the new module's `.tsv` file. The -i flag is the upstream `bcodmont-edit.owl` application ontology. There are multiple `--prefix` flags including for `RO`, `BSM` and any imported ontologies, for example `PCO`. Make sure to add prefix flags for all imported ontologies. The `--ontology-iri` flag is the machine-readable name of the owl module generated, it is of the format `https...prefix.../BCODMO_SM/module/robot_templates/sub-module.owl`. Finally the `-o` flag is the output file, and should be of the format`module/robot_templates/sub-module.owl`. Make sure to run this new command from `bcodmont/src/ontology/BCODMO_SM`, debugging as needed till it compiles and produces the expected output file, e.g. `biology/robot_templates/ecology.owl`. 

**6.3\)** Prepare robot merge command\(s\), see and update the doccumentation in the [Merge Robot templates](https://github.com/BCODMO/bcodmont/blob/main/docs/technical-documentation.md#3-merge-robot-templates) section. Setting up this next section will depend on wether or not you keep the existing axioms from the imported ontologies. At a high level there may be some assertion made in the upstream import ontologies that we don't want to keep in BCO-SM, to remove complexity or change things slightly to have a simplified hierarchy in BCO-SM. 

**6.3 a\)** In the case where **do not** remove axioms, aka keep the import ontologies as they are, we'll need to modify the command in the [Merge Imports and Preliminary Robot Templates](https://github.com/BCODMO/bcodmont/blob/main/docs/technical-documentation.md#merge-imports-and-preliminary-robot-templates) section to add 1\) any new imports and 2\) the new module file. In the case of adding the new `ecology` module we'd need to add both the PCO import:

```text
--input ../imports/pco_import.owl
```

As well as the new module file:

```text
--input biology/robot_templates/ecology.owl
```

To the existing command. 

**6.3 b\)** In the case where **do** want to remove axioms, aka modify the import ontologies hierarchy, we'll need to, for each new import ontology add a new command to the [Make Object Property \(OP\) Free Versions of Select Import Ontologies](https://github.com/BCODMO/bcodmont/blob/main/docs/technical-documentation.md#make-object-property-op-free-versions-of-select-import-ontologies) section. For example in the case of the new `ecology` module we'd add the following to make an hierarchy free version of PCO:

**Make OP free version of PCO import**. Only run after adding to PCO import.

```text
robot remove --input ../imports/pco_import.owl --axioms logical annotate --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/intermediate/pco_import_axioms_removed.owl" --version-iri "http://purl.obolibrary.org/BCODMO_SM/intermediate/pco_import_axioms_removed.owl" --output intermediate/pco_import_axioms_removed.owl
```

The above command can be modified to switch all instances of "pco" for the relevant import ontology or ontologies codes. The code is the same as that used in the ontology import. 

Also modify the command in the [**Merge Modules with Axiom free Import Ontologies**](https://github.com/BCODMO/bcodmont/blob/main/docs/technical-documentation.md#merge-modules-with-axiom-free-import-ontologies) section to include the outputs from the previous step. For example for PCO, we'd need to add a new input flag to the command \(for the file created in the step above\):

```text
--input intermediate/pco_import_axioms_removed.owl
```

Note that making the axiom removed version of the ontology might not always be necessary and one could instead merge the regular import as is the case for `iao_import.owl`. //TODO double check and remove this if necessary. 

**Step 6.4\)** Next we'll need to add a robot filter command to the [**Filter merge product to create final modules**](https://github.com/BCODMO/bcodmont/blob/main/docs/technical-documentation.md#4-filter-merge-product-to-create-final-modules) section. Make sure to do so under the correct top level module. For example for adding the new `ecology`module under biology we'd add the following:  


**6.4 a\)** If following step **6.3 a**, the command would be:

Filter **ecology** from the regular merged ontology

```text
robot filter --input merge_products/BCODMO_SM_merged.owl --prefix "bsm:http://bcodmo/sm#" --select "oboInOwl:inSubset=bsm:ecology" --select annotations  --signature true annotate --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/biology/ecology.owl" --version-iri "http://purl.obolibrary.org/BCODMO_SM/biology/ecology.owl" --output biology/ecology.owl
```

**6.4 b\)** If following step **6.3 b**, the command would be:

Filter **ecology** from the axiom-free merged ontology

```text
robot filter --input merge_products/BCODMO_SM_axioms_removed_merged.owl --prefix "bsm:http://bcodmo/sm#" --select "oboInOwl:inSubset=bsm:ecology" --select annotations --signature true annotate --ontology-iri "http://purl.obolibrary.org/BCODMO_SM/biology/ecology.owl" --version-iri "http://purl.obolibrary.org/BCODMO_SM/biology/ecology.owl" --output biology/ecology.owl
```

In either case it's important to update the `--select "oboInOwl:inSubset=bsm:ecology"` to be the name of the new module and the same url as added in **step 4**. These steps will generate the final version of the newly added module. In the case of `ecology`, `biology/ecology.owl`.

\*\*\*\*

### Setting up BCODMONT on a new computer

BCODMONT is based on the ODK and therefore requires [docker](https://www.docker.com/).

**Step 1\)** On the new system install docker \(if not already installed\), See the [https://www.docker.com/get-started](https://www.docker.com/get-started). Docker desktop on a personal computer, \(docker hub for a cloud image?\).

**Step 2\)** Following the instructions described in the [ODK](https://github.com/INCATools/ontology-development-kit), pull the ODK docker image

```text
docker pull obolibrary/odkfull
```

**Step 3\)** Clone BCODMONT repository  
Note might suggest to do via ssh instead of HTTPS and using ssh keys as github is deprecating passwords as of the summer 2021.

```text
git clone https://github.com/BCODMO/bcodmont.git
```

## Future Sections

### Configuring and testing memory for running imports

Temp notes on using CHEBI import I tried the following in the `run.sh`script to run chebi import but it still failed. Problem is documented [here](https://github.com/INCATools/ontology-development-kit/blob/master/docs/DealWithLargeOntologies.md). I think min 8 GB is required for CHEBI.

```text
docker run -m 4g -v $PWD/../../:/work -w /work/src/ontology -e ROBOT_JAVA_ARGS='-Xmx3G' -e JAVA_OPTS='' --rm -ti obolibrary/odkfull "$@"
```

Try again with: 8 GB instance and the following \(if the regular command doesn't work\).

```text
docker run -m 8g -v $PWD/../../:/work -w /work/src/ontology -e ROBOT_JAVA_ARGS='-Xmx7' -e JAVA_OPTS='' --rm -ti obolibrary/odkfull "$@"
```

\*\*\*\*

**//TODO section about** using wget or similar to download robot templates via the command line:

#### Biology

//TODO: Re-download **anatomy**.tsv:

```text
//TODO
```

Try from [https://angelov.ai/post/2020/wget-files-from-gdrive/](https://angelov.ai/post/2020/wget-files-from-gdrive/) or [https://silicondales.com/tutorials/g-suite/how-to-wget-files-from-google-drive/](https://silicondales.com/tutorials/g-suite/how-to-wget-files-from-google-drive/)or or [https://bcrf.biochem.wisc.edu/2021/02/05/download-google-drive-files-using-wget/](https://bcrf.biochem.wisc.edu/2021/02/05/download-google-drive-files-using-wget/) or similar or the command like Nico used:

```text
MIXTURES="https://docs.google.com/spreadsheets/d/e/2PACX-1vQr0nBpt5ySurXeTgAIoiI2t1wC3xk2fGrZRVQOvc_ugziNlKHoNj4-xbVGnz84ohOVc-0g693LWiiz/pub?output=tsv"

templates/mixtures.tsv:
    wget $(MIXTURES) -O $@
```

...

@Kai - not sure if this is what you're looking for, but rclone is another program that easily lets you move files between googledrive & machines vis command line/programmatic interface

\*\*\*\*

**//TODO section about** Version control how to release a new version. Can release a V1 once all mods are in place but want to have another section. Directions on how do releases with named versions.

**//TODO section about** Later once we move away form google sheets as source of truth for robot templates, we'll need some directions on how to edit them there. How do this with branches and pull requests rather than just commits to master.

