# Technical documentation

## Overview

BCODMONT is an OBO style application ontology created using the [Ontology Development Kit \(ODK\)](https://github.com/INCATools/ontology-development-kit). As such it makes use of a variety of features generated by ODK such as ontology imports a standardized makefile and githhub release-generation. 

The **BCO**DMO **S**emantic **M**odel \(**BCO-SM**\) is created within BCODMONT and acts as a controlled vocabulary for the BCODMO data managers to annotate data with. 

The BCO-SM is loosely based on the [NERC Vocabulary Server \(NERC-VS\)](https://www.bodc.ac.uk/resources/products/web_services/vocab/) model, but draws from terminology from various ontologies from the [Open Biological and Biomedical Ontology \(OBO\) Foundry](http://www.obofoundry.org/). It can be though of loosely as a hybrid of the two models, bridging them by playing to the strengths of both systems. 

### Modules:

BCO-SM has the following domain specific modules which can be used together in combination to annotate datasets. 

**Figure 1**: Overview of the BCO-SM 

![](.gitbook/assets/image%20%282%29.png)

#### Biology

The Biology module consists of several sub-modules [Anatomy](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/biology/anatomy.owl),[ Physiology](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/biology/physiology.owl), Ecology, Biomolecules, and Organism. It is intended to represent the biologically relevant information pertinant to datasets. 

#### Chemistry

The Chemistry module consists of two sub-modules [Compound](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/chemistry/compound.owl) and [Element](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/chemistry/element.owl). The Compound sub-module consists of a set of chemicals that are determined by analysis focusing on a multi-atomic compound, e.g, `nitrate` or `carbohydrate`. Whereas the Element sub-module consists of chemicals that are determined by analysis focusing on specific elements including isotopes and valence states, e.g. `elemental Nitrogen` or `Nitrogen-15`.

#### Matrix

The Matrix module represents the environmental context from which datasets were collected. It is based on both the NERC Matrix vocabulary as well as the [MIxS](https://gensc.org/mixs/) \(Minimum Informtion about any "x" Sequence\) standards from the Genomic Standards Consortium \(GSC\). BCO-SM's Matrix is divided into four sub-modules which like the MIxS model represent the environmental context data at increasing levels of granularity. The first module [Region](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/matrix/region.owl), describes at the highest level if data were collected from a `polar`, `temperate` or `tropical` region. 

The next three sub-modules draws from terminology from the Environment Ontology \(ENVO\); following the MIxS ENVO annotation framework which includes the use of \(one or more terms\) from each of the, [Biome](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/matrix/biome.owl), [Context](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/matrix/context.owl), and [Material](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/matrix/material.owl) sub-modules. These are analogous to the `broad-scale environmental context`, `local environmental context`, and `environmental medium`, fields used in the MIxS model. In the **Biome** module are terms by which to report which major environmental system from which a sample came, namely an `aquatic`, `marine`, or `terrestrial` biome. `Aquatic`referring to both freshwater as well as non-marine saline environments, e.g., saline lakes. 

The **Context** sub-module contains terms by which to describe the local vicinity of a sample which may have a causal influences on data, e.g. if the data were collected from an `esturay` or a  water layer such as the `marine benthic zone`. 

Finally the **Material** sub-module contains terms for environmental or other material or materials that immediately surrounded or comprise samples e.g., `sea water`, `plankton`, or `sediment`.

#### Operational

The [Operational](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/operational/operational.owl) module describes information about sample processing e.g., `filter fractionation size`, sample post-processing e.g.,  `external repository` as well as spatiotemporal sample collection information e.g., `latitude` or `collection date`.

#### Physics

The Physics module is split up into two sub-modules. The first of which, [Characteristic](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/physics/characteristic.owl) consists of terms describing qualities, characteristics or attributes of materials \(both biotic and abiotic\), e.g., the `depth` or `salinity` measured in sea water, or the `color` used to describe an organism or water body. The [Phenomenon](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/physics/phenomenon.owl) module consists of physical phenomena either measured or believed to have a causal influence on a sample, e.g., `Taylor column` `deep ocean current` or `thunderstorm`. Note at the moment there are some overlapping terms between the Physics Phenomenon and Matrix Context modules this is intential but we may want to contstrain terms like `epilimnion` to only Matrix.  

#### Quantifiers

The [Quantifiers](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/quantifiers/quantifiers.owl) module consists of terms to describe quantitative measurements conducted on samples within datasets, including classes like `concentration`, `length`, or `standard deviation`. Which are to be used in conjunction with terms from other modules. 

### OBO Ontologies Used

Cross reference the [OBO Foundry hompage](http://www.obofoundry.org/).

**The Basic Formal Ontology \(BFO\)**: is the upper level ontology upon which OBO Foundry ontologies are built. BCO-SM makes use of BFO terms like `site`. Browse BFO from [here](https://www.ebi.ac.uk/ols/ontologies/bfo).

**The Chemical Entities of Biological Interest \(CHEBI\)**: is a structured classification of molecular entities of biological interest focusing on 'small' chemical compounds. CHEBI terms comprise most of the BCO-SM chemistry terms e.g. `nitrate` and are also used in other BCO-SM modules such as the Biomolecules submodule from Biology for terms like `DNA` or `RNA`. Browse CHEBI from [here](https://www.ebi.ac.uk/ols/ontologies/chebi). 

**The Cell Ontology \(CL\)**: is a structured controlled vocabulary for cell types in animals. BCO-SM makes use of CL terms, e.g. `red blood cell` within the Biology Anatomy module. Browse CL from [here](https://www.ebi.ac.uk/ols/ontologies/CL).

**The Core Ontology for Biology and Biomedicine \(COB\)**: brings together key terms from a wide range of OBO projects to improve interoperability, attempting to standardize terms important across disciplines and ontologies. BCO-SM when possible tries to make use of COB terms within Biology modules, e.g. `organism`, or `cell`, which might also be found in other OBO ontologies. Browse COB from [here](https://www.ebi.ac.uk/ols/ontologies/cob).

**The Environment Ontology \(ENVO\)**: describes environmental features, biomes, ecosystems, planetary or other astronomical bodies, their parts, or environmental processes. The BCO-SM primarily uses ENVO terms, \(e.g.,  `abyssal plain` or `first year ice` within the Matrix modules to comprehensively describe the environmental context of collected samples, at various levels of granularity, following the MIxS model used by genomic repositories such as NCBI and EBI.  BCO-SM also uses ENVO terms in the Physics Phenomenon and Biology Ecology modules. Browse ENVO from [here](https://www.ebi.ac.uk/ols/ontologies/envo).

**The Gene Ontology \(GO\)**: describes the function of genes and gene products. BCO-SM uses GO terms, e.g., `photosynthesis` and `repsiration` in the Biology Physiology submodule. Browse GO from [here](https://www.ebi.ac.uk/ols/ontologies/go).

**The Information Artifact Ontology \(IAO\)**: sometimes refereed to as the OBO Metadata Ontology \(OMO\), is the OBO ontology for information entities. IAO includies both classes, e.g., `data set` or `setting datum` and annotation properties, e.g. `defintion`. When Possible IAO classes are used within the BCO-SM Operational module. Browse IAO from [here](https://www.ebi.ac.uk/ols/ontologies/iao).

**The Mass spectrometry ontology \(MS\)**: is a structured controlled vocabulary for the annotation of experiments concerned with proteomics mass spectrometry. Browse MS from [here](https://www.ebi.ac.uk/ols/ontologies/ms).

**The Ontology for Biomedical Investigations \(OBI\)**: is an integrated ontology for the description of life-science and clinical investigations. BCO-SM makes use of various OBI classes, e.g.,  `sample` and `longitude` within the BCO-SM Operational module. Browse OBI from [here](https://www.ebi.ac.uk/ols/ontologies/obi).

**The Phenotype And Trait Ontology \(PATO\)**: is an ontology of phenotypic qualities, aka, properties, attributes or characteristics. These include both biological attributes of organisms, e.g. `size` as well as physical qualities such as `salinity`, or `acidity`, or `color`. BCO-SM uses PATO in both the Quantifiers and Physics Characteristic modules. Browse PATO from [here](https://www.ebi.ac.uk/ols/ontologies/pato).

**The Population and Community Ontology \(PCO\)**: is an ontology about groups of interacting organisms such as populations and communities. PCO terms such as `Simpson diversity index`,  and `community` are used in the Biology Ecology module. Browse PCO from [here](https://www.ebi.ac.uk/ols/ontologies/pco).

**The Relation Ontology \(RO\)**: is a controlled set of relationship types shared across OBO ontologies in order to enable interoperability across ontologies. The BCO-SM module currently do not heavily leverage RO relationships, however future iterations of the system could to enable more advanced querying. Browse RO from [here](https://www.ebi.ac.uk/ols/ontologies/ro).

**The Statistical Methods Ontology \(STATO\)**: is a general-purpose statistics Ontology, providing coverage for processes such as statistical tests, their conditions of application, and information needed or resulting from statistical methods. BCO-SM leverages STATO terms such as `median` ,  or `confidence interval` in the Quantifiers module. Browse STATO from [here](https://www.ebi.ac.uk/ols/ontologies/stato).

**The multi-species anatomy ontology \(UBERON\)**: is an integrated cross-species anatomy ontology covering animals and bridging multiple species-specific ontologies. BCO-SM uses UBERON terms such as `blood`, or `juvenile` in Biology modules such as Anatomy and Physiology respectively. Browse UBERON from [here](https://www.ebi.ac.uk/ols/ontologies/uberon).

**The Sequence types and features ontology \(SO\)**: is a structured controlled vocabulary for sequence annotation, for the exchange of annotation data and for the description of sequence objects in databases. Browse SO from [here](https://www.ebi.ac.uk/ols/ontologies/SO). //TODO will need to add this as new import \(can use as example on how to add new ones\). 

### Other Ontologies Referenced but not directly used

CHMO //TODO might want to use this.

EDAM

GENEPIO

NCIT

NMR

OMIABIS

PRIDE

SIO //TODO fix in biomolecules to use BSM instead.

## New Term Requests

In the interim stages of the BCO-SM's development we expect there to be many changes made to the model, mostly additions but perhaps also some term replacements or reshuffeling. As such this section serves as a guide for data managers to request new terms to be added to BCO-SM. Once the model is a bit more solidified and the downstream data processing in order, the  process of adding new terms could be taken up by interested Data Mangers. 

For the time being, data managers can do the following to request new terms. 

1\) First navigate to the [bcodmont/issues](https://github.com/BCODMO/bcodmont/issues/new) page. 

2\) Create a new issue describing the requested term. Explain what it is in clear concise language what the missing term is and what BCO-SM module you think it should be added to. Make sure for now to tag @kaiiam in the github issue \(untill the model is stablized and others are trained to handle this as well or in place\). 

3\) Make sure to check throughly check through the existing modules in [https://github.com/BCODMO/bcodmont/tree/main/src/ontology/BCODMO\_SM](https://github.com/BCODMO/bcodmont/tree/main/src/ontology/BCODMO_SM). //TODO either add link to web protege versions or other way to visualize these such as the EBI-docker. 

4\) **Optional** Search for appropriate existing terms. Note that when possible BCO-SM uses existing OBO terms to both have greater interoperability with external projects as well as reuse the work of countless people who have contributed to many well-grounded biological and biomedical ontologies. To explore existing OBO ontology terms navigate to the [OLS ontology lookup page](https://www.ebi.ac.uk/ols/index) and type in your term into the search bar. There are many OBO ontologies not all of which are of the same level of quality, therefore it can be a bit confusing to search. See the `OBO Ontologies Used` section of `Overview` for a list of the OBO ontologies we are using for BCO-SM. This list is by no means the only OBO ontologies to search for terms in, but it represents a very good place to start. 

5\) **Optional** if you want to go the extra mile check out the following section on perparing new BCO-SM terms for the step by step detailed instructions for adding to modules via the Robot templates by which they are compiled. 

### Preparing new BCO-SM terms

The working versions of the BCO-SM robot templates are currently the google drive sheets \(linked within the _BCO-DMO Semantic Data Project_ trello page. At the moment these are exported down to the [BCODMO\_SM](https://github.com/BCODMO/bcodmont/tree/main/src/ontology/BCODMO_SM) sub-directories e.g., [biology/robot\_templates/anatomy.tsv](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/biology/robot_templates/anatomy.tsv).  However, relying on google sheets as the source of truth better should not to be considered a long term solution. At some point once we've stabilized the system, we will switch to treating the `tsv` files within the `X_module/robot_templates/` to being the source of truth \(place to work from\).

#### Basics about Robot templates

Robot templates are structured as follows, the first row contains `headers`, these are human-readable labels that are not complied by robot. These can be renamed as needed. The second row contains the `templates` which must contain strings formatted as specified in the [Robot template documentation](https://robot.obolibrary.org/template). Please do not change these. All subsequent rows are for the addition of new terms or the modification of existing ones. 

#### BCO-SM Category Headers

At the moment the first several columns contain headers such as `Sub category 1:`. These columns are not complied by the robot templates but are used to visualize the basic parent child hierarchy that is complied when the template is converted to owl. These are not strictly necessary \(and could be removed, however, they can be helpful when adding new terms to the model\). 

#### Ontology IDs

In order for a term \(a row in the spreadsheet\) to be recognized by robot and compiled into owl code, it must have an entry in the `ontology ID` column. This must either include a BCODMO Semantic Model \(BCO-SM\) ID, or an ID from another OBO ontology. BCO-SM terms must start with the prefix `BSM:` and be followed by seven digits e.g., `BSM:0200001`. BCO-SM IDs must be within an appropriate numeric ID range as specified within the [BCODMO\_SM/idranges.owl](https://github.com/BCODMO/bcodmont/blob/main/src/ontology/BCODMO_SM/idranges.owl) file. 

Imported terms from other OBO ontologies such as ENVO, UBERON, GO etc must follow this same formatting, e.g, with imported ENVO terms one must use the prefix `ENVO:` followed by the correct identifier number, e.g. `ENVO:03000102`. This is refereed to as the `curie` ID formatting style. 

#### Labels

In the `label` column, put the primary labels for new or imported terms, e.g., `marsh` or `ice`. BCO-SM follows lower case naming conventions, with exceptions for proper nouns, e.g., `Taylor column` or `WMO blizzard`. In some instances BCO-SM renames terms imported from existing OBO ontologies in order to simplify them. For example, we rename `ENVO:01000811` `nitrogen-oxygen planetary atmosphere` to `atmosphere`.

#### Parent classes

BCO-SM terms and imported terms should make use of the `parent class` relationship. For example if you were adding a new class `ice fog`, you'd add the `parent class` as `fog`.  You can simply type or copy paste the label of an existing term into the `parent class` column.

To discover new OBO terms which might be relevant to use as parent classes, you can browse them from the ENVO from the [EMBL-EBI Ontology Lookup Service](https://www.ebi.ac.uk/ols/ontologies/) \(OLS\). The OLS search can be filtered by ontology, e.g., [https://www.ebi.ac.uk/ols/ontologies/chebi](https://www.ebi.ac.uk/ols/ontologies/chebi), or [https://www.ebi.ac.uk/ols/ontologies/go](https://www.ebi.ac.uk/ols/ontologies/go) etc. 

Additional documentation about how to browse the Environment Ontology \(ENVO\), as well as its contents, can be found at the [navigating ENVO](https://github.com/EnvironmentOntology/envo/wiki/Navigating-ENVO) github wiki page. The navigation page describes how to go about viewing ENVO \(or other OBO ontologies\), as well as what are main hierarchies within ENVO, giving examples of each.

#### Matches

BCO-SM makes use of the SKOS `exact match` and `close match` annotation properties. `exact match` should be used in cases where two semantics are intended to be the same. For example if an OBO term is found or created that can replace a BCO-SM term.`close match` can be used in cases where two semantics maybe the similar but aren't exactly the same representation. For example terms form the [NCI Thesaurus OBO Edition](https://www.ebi.ac.uk/ols/ontologies/ncit) e.g., `NCIT:C12508`.

#### Definitions

Although definitions and other annotation properties are included for imported OBO terms, new BCO-SM terms, are envisioned to eventually be added to appropriate OBO Foundry Ontologies. As such we encourage new BCO-SM terms to make use of OBO conventions for writing definitions. This includes advise people to follow a genus-differentia structure, i.e., `An A is a B that C's`. The genus refers the parent term, and the differentia is that which makes a term different \(or more specific\) than its parent. Doing this also helps to identify parent/child relationships between terms.

For example in ENVO, example they define [land ice mass](http://purl.obolibrary.org/obo/ENVO_01001547) as:

> An ice mass which has formed over land.

Another example, in UBERON, they define [tissue](http://purl.obolibrary.org/obo/UBERON_0000479) as:

> \[A\] multicellular anatomical structure that consists of many cells of one or a few types, arranged in an extracellular matrix such that their long-range organisation is at least partly a repetition of their short-range organisation.

Using a genus-differentia structure helps both to formulate a concise and simple definition as well as determine the appropriate parent class. In cases where there are multiple differentiating factors, they can be written out as a list, for example:

> A blizzard which has 1\) winds of 35 miles per hour \(approximately 56 kilometers per hour\), 2\) lasts for a period of at least three hours, and 3\) occurs in extremely cold temperatures.

For more information about writing ontology definitions please see [this blog post](https://douroucouli.wordpress.com/2019/07/08/ontotip-write-simple-concise-clear-operational-textual-definitions/) written by Dr. Chris Mungall.

#### Comments

In the `comment` column add any supporting information which is relevant to the term but not essential for its definition. As described in the above [blog post](https://douroucouli.wordpress.com/2019/07/08/ontotip-write-simple-concise-clear-operational-textual-definitions/), any information which gets cut out while trying to write a concise genus-differentia style definition can be put into the comment.

#### Editors notes

The `editors note` column serves as a place for ontology developers to make notes about how to better engineer the term or discuss future developments. For new BCO-SM terms this is where one can suggest an an appropriate OBO ontology and possible parent class for such a term.

#### Synonyms

Following OBO common practices, the BCO-SM robot templates contains four types of synonyms: `exact synonym`, `broad synonym`, `narrow synonym`, `related synonym`. These follow SKOS style \(broader or narrow\) conventions. 

Use `exact synonym` for interchangeable term labels, for example `seawater` and `sea water`. 

Use `broad synonym` when a synonym might refer to more than just the term of interest, for example the term `acid rainfall` has `acid rain` as a broad synonym because `acid rain` might refer to more than just the process of acid rainfall. 

When a synonym refers to something more specific than the class label use `narrow synonym`, for example `road` has narrow synonym `highway`. 

If you unsure about a synonym being broader or narrower then use `related synonym`.

#### Cross references

The `cross reference` column provides a generic way to link to other resources such as mappings to other vocabularies. This could for example be used to map BCO-SM terms to \(individual\) terms from the [NERC vocabularies](https://www.bodc.ac.uk/resources/products/web_services/vocab/).

#### BCODMO master parameter

The `BCODMO master parameter` column is used to link BCO-SM terms to the \(previous system the\) BCODMO master parameters, e.g., `http://lod.bco-dmo.org/id/parameter/1982`.

#### Subsets

BCO-SM has both `category subset` and `Entity subset` columns. At the moment the `Entity subset` column is not being used, however the `category subset` is essential for the BCO-SM compilation process. As such the `category subset` must include a module-specific iri such as `http://bcodmo/sm#material` or `http://bcodmo/sm#anatomy`.

#### Columns with multiple entries

Any columns with `SPLIT=|` at the end of the string in the template \(second\) row allow for multiple `|` delimited inputs where each of which is compiled separately. For example adding `boulbie|buran|purga` within the `narrow synonym` column, which has the template `AL oboInOwl:hasNarrowSynonym@en SPLIT=|` will produce three separate narrow synonym annotation properties:

![image](https://user-images.githubusercontent.com/12255688/100488568-008bce00-30dd-11eb-8c83-45324df24f0f.png)

**//TODO section** after the \`New Term Requests Section, about adding new terms and compiling them. Have this be part of the docs walkthrough on the 26th. Talk about when to commit i.e, with a set of new terms to a module. 

## Maintaining BCO-SM

### Managing Imports

In order to re-use ontology terms form existing OBO ontologies  we need to import them into our repository. Currently BCO-SM makes use of various OBO ontologies, see the `OBO Ontologies Used` section. 

#### Add new terms to import

In order to import new terms from ontologies we are already importing from we need to do the following:

#### Running all imports

We can run all existing imports from `bcodmont/src/ontology` with the following command: 

```text
./run.sh make all_imports
```

Note when running on linux systems depending on permissions one may need to add `sudo` in front of the above command. 

#### Running individual imports

To re-run an individual ontology \(instead of all imported ontologies\) one can run a command such as the following, where for example the Gene Ontology `go` is the ontology of interest. One would run the following from `bcodmont/src/ontology`: \(note `sudo` maybe required\)

```text
./run.sh make imports/go_import.owl
```



### Import A New Ontology

In order to import a new OBO ontology we will need to do the following steps. Documented in this example is the process by which to import the **Sequence types and features ontology \(SO\)** into BCODMONT. This process will work with other OBO foundry ontologies which are available from [`http://purl.obolibrary.org/obo/`](http://purl.obolibrary.org/obo/bfo.owl) followed a valid ontology file e.g. `bfo.owl`. 

In the following directory `bcodmont/src/ontology`is the ontology `Makefile`. The `Makefile` is the heart of the BCODMONT application ontology created by the Ontology Development Kit \(ODK\) in the initial release. The `Makefile` contains instructions for managing a variety of tasks including managing ontology imports, and releases. Within the `Makefile` \(on approximately line 75 at the time of creating these docs\), there will be the `IMPORTS` code-block which will resemble the following:

```text
IMPORTS = ro pato envo iao obi uo uberon go stato ms bfo cl cob pco chebi

IMPORT_ROOTS = $(patsubst %, imports/%_import, $(IMPORTS))
IMPORT_OWL_FILES = $(foreach n,$(IMPORT_ROOTS), $(n).owl)
IMPORT_FILES = $(IMPORT_OWL_FILES)
```

**Step 1\)** Add the new target ontology to be imported, \(e.g., **`so`**\)  to the list of `IMPORTS`, e.g. `IMPORTS = ro pato envo iao obi uo uberon go stato ms bfo cl cob pco chebi so`. Make sure to add an appropriate commit message such as:

```text
git commit Makefile -m 'Add so import to import target list'
```

**Step 2\)** In the `Mirroring upstream ontologies` section:

```text
# ----------------------------------------
# Mirroring upstream ontologies
# ----------------------------------------
#

IMP=true # Global parameter to bypass import generation
MIR=true # Global parameter to bypass mirror generation

## ONTOLOGY: ro
## Copy of ro is re-downloaded whenever source changes
mirror/ro.trigger: $(SRC)

mirror/ro.owl: mirror/ro.trigger
	@if [ $(MIR) = true ] && [ $(IMP) = true ]; then $(ROBOT) convert -I $(URIBASE)/ro.owl -o $@.tmp.owl && mv $@.tmp.owl $@; fi

.PRECIOUS: mirror/%.owl
```

There are a series of make targets for various ontologies e.g., `ro` shown above. Copy and paste the following at the end of the section, replacing `xyz` with the target ontology name in lower case, e.g., `so`. 

```text
## ONTOLOGY: xyz
## Copy of xyz is re-downloaded whenever source changes
mirror/xyz.trigger: $(SRC)

mirror/xyz.owl: mirror/xyz.trigger
	@if [ $(MIR) = true ] && [ $(IMP) = true ]; then $(ROBOT) convert -I $(URIBASE)/xyz.owl -o $@.tmp.owl && mv $@.tmp.owl $@; fi

.PRECIOUS: mirror/%.owl
```

Make sure to add an appropriate commit message such as:

```text
git commit Makefile -m 'Add SO import mirror lines'
```

**Step 3\)** Add import terms file. 

In the directory `bcodmont/src/ontology/imports` create a new `xyz_terms.txt` file where `xyz` is replaced by the new ontology name, e.g., `so`. This could be done by copying one of the existing files, e.g., `cp bfo_terms.txt xyz_terms.txt`. For SO the command is 

```text
cp bfo_terms.txt so_terms.txt
```

**Step 4\)** Add new term\(s\) to import to the `xyz_terms.txt` file. 

In our example we'll add to the following to be the content of the `so_terms.txt` file using any standard text editor such as atom, vim, nano, emacs, gedit, notepad++, textedit, etc depending on what system you're using. 

```text
### BCODMO_SM Biology/Biomolecules terms:
SO:0001031 #reverse sequence
```

Note the following, `#`'s are comments which are not compiled. Hence line 1 is not strictly necessary but it helps to keep track of where imported terms were intended to be used. Once imported new terms do not need to be duplicated, i.e., we only ever need to specify `SO:0001031`once in the `so_terms.txt` file \(even if it's used in multiple modules\). Imported terms must follow this format referred to as CURIE format with the uppercase ontology name, e.g. `SO`, followed by the numeric ID `0001031`. Atfer the curie it's typical but not stricly required to put the term label e.g., `#reverse sequence`.

Make sure to **Add**  the new .txt file: 

```text
git add so_terms.txt
```

As well as commit it with an appropriate commit message such as:

```text
git commit so_terms.txt -m 'Add term imports file for SO'
```

**Step 5\)** Run make imports:

This could either be done with the generic command to run all imports:

```text
./run.sh make all_imports
```

Or by running an individual import e.g.,

```text
./run.sh make imports/so_import.owl
```

Note that on linux it might be necessary to run with sudo `sudo ./run.sh make imports/so_import.owl`

This will produce the so import owl file containing desired classes which can then be used in the BCO-SM. Make sure to add the new import ontology file

```text
git add imports/so_import.owl
```

As well as commit it with an appropriate message such as:

```text
git commit imports/so_import.owl -m 'Run make all_imports'
```

\*\*\*\*

### **Adding BSM Modules**

**//TODO section about** adding a new BCO-SM module use one of the biology mods as an example. 

\*\*\*\*



## Setting up BCODMONT on a new computer

BCODMONT is based on the ODK and therefore requires [docker](https://www.docker.com/). 

**Step 1\)** On the new system install docker \(if not already installed\), See the [https://www.docker.com/get-started](https://www.docker.com/get-started). Docker desktop on a personal computer, \(docker hub for a cloud image?\). 

**Step 2\)** Following the instructions described in the [ODK](https://github.com/INCATools/ontology-development-kit), pull the ODK docker image

```text
docker pull obolibrary/odkfull
```

**Step 3\)** Clone BCODMONT repository  
Note might suggest to do via ssh instead of HTTPS and using ssh keys as github is deprecating passwords as of the summer 2021.

```text
git clone https://github.com/BCODMO/bcodmont.git
```



## Future Sections

### Configuring and testing memory for running imports

Temp notes on using CHEBI import I tried the following in the `run.sh`script to run chebi import but it still failed. Problem is documented [here](https://github.com/INCATools/ontology-development-kit/blob/master/docs/DealWithLargeOntologies.md).  I think min 8 GB is required for CHEBI. 

```text
docker run -m 4g -v $PWD/../../:/work -w /work/src/ontology -e ROBOT_JAVA_ARGS='-Xmx3G' -e JAVA_OPTS='' --rm -ti obolibrary/odkfull "$@"
```

Try again with: 8 GB instance and the following \(if the regular command doesn't work\). 

```text
docker run -m 8g -v $PWD/../../:/work -w /work/src/ontology -e ROBOT_JAVA_ARGS='-Xmx7' -e JAVA_OPTS='' --rm -ti obolibrary/odkfull "$@"
```

**//TODO section about** Version control how to release a new version. Can release a V1 once all mods are in place but want to have another section. Directions on how do  releases with named versions. 

**//TODO section about** Later once we move away form google sheets as source of truth for robot templates, we'll need some directions on how to edit them there. How do this with branches and pull requests rather than just commits to master. 

